src/compiler/scala/tools/nsc/typechecker/Namers.scala

```diff
@@ -159,11 +159,15 @@ trait Namers extends MethodSynthesis {
       setPrivateWithin(tree, sym, tree.mods)
 
     def inConstructorFlag: Long = {
-      val termOwnedContexts: List[Context] =
-        context.enclosingContextChain.takeWhile(c => c.owner.isTerm && !c.owner.isAnonymousFunction)
-      val constructorNonSuffix = termOwnedContexts exists (c => c.owner.isConstructor && !c.inConstructorSuffix)
-      val earlyInit            = termOwnedContexts exists (_.owner.isEarlyInitialized)
-      if (constructorNonSuffix || earlyInit) INCONSTRUCTOR else 0L
+      @tailrec def go(contextChain: List[Context]): Long = contextChain match {
```

Comment:
#### Small efficiency improvement

The current code allocates a whole list with the call to `enclosingContextChain`, and then iterates through that list. Perhaps we could cut down that middle list by using the code of `enclosingContextChain`: 
```Scala
@tailrec def go(c: Context): Long = c match {
  case NoContext => 0L // corresponds to the nil case
  /* [...] rest of cases here */
  case _ => go(c.outer)
}
``` 

Nevertheless, IMHO this PR is good to merge.

Reply:
Good point :+1: I think in this case if-else will be cleaner than a pattern match.